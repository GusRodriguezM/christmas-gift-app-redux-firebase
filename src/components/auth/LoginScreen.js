import React, { useMemo, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { startGoogleSignIn, startLoginWithEmailPassword } from '../../store/slices/auth';
import { removeErrorMessage, setErrorMessage } from '../../store/slices/ui';
import { useForm } from '../../hooks';

import { Button } from '../styles/shared/Button.styled';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import Toast from '../styles/Toast/Toast';

export const LoginScreen = () => {

    const { status } = useSelector( state => state.auth );
    const { msgError } = useSelector( state => state.ui );
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const isAuthenticating = useMemo(() => status === 'checking', [status]);

    useEffect(() => {
        if(msgError.length > 0){
            Toast.fire({
                icon: 'error',
                title: `${msgError}`
            });
        }
    }, [msgError]);

    const isFormValid = () => {
        if(validator.isEmpty(email) && validator.isEmpty(password)){
            dispatch( setErrorMessage('Email and password are required') );
            return false;
        }else if(validator.isEmpty(email)){
            dispatch( setErrorMessage('The email is required') );
            return false;
        }else if(!validator.isEmail(email)){
            dispatch( setErrorMessage('The email is not valid') );
            return false;
        }else if(validator.isEmpty(password)){
            dispatch( setErrorMessage('The password is required') );
            return false;
        }

        dispatch( removeErrorMessage() );
        return true;
    }

    const handleLogin = (e) => {
        e.preventDefault();
        
        if(isFormValid())
            dispatch( startLoginWithEmailPassword({email, password}) );
    }

    const handleGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
    }

    return (
        <Container>

            <h1>Login</h1>

            <Container.AuthForm onSubmit={handleLogin}>

                <Input
                    id='name'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button 
                    type='submit'
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Login</span>
                    <i className="fa-solid fa-user-lock"></i>
                </Button>

                <Button 
                    onClick={handleGoogleSignIn}
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Google Login</span>
                    <i className="fa-brands fa-google"></i>
                </Button>

            </Container.AuthForm>

            <Link
                aria-disabled={isAuthenticating}
                to='/auth/register'
            >
                Create a new account
            </Link>
        </Container>
    )
}