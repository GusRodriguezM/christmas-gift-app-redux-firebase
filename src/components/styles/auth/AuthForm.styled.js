import styled from 'styled-components';

const AuthForm = styled.form`
    width: 100%;
    height: 60%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-evenly;
`;

export default AuthForm;