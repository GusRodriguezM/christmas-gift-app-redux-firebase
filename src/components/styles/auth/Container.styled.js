import styled from 'styled-components';

import AuthForm from './AuthForm.styled';

const Container = styled.div`
    width: 450px;
    height: 450px;
    background: rgba( 211, 226, 236, 0.5 );
    box-shadow: 0 8px 32px 0 rgba( 31, 38, 135, 0.37 );
    backdrop-filter: blur(4px);
    border-radius: 10px;
    display: flex;
    flex-direction: column;
    align-items: center;
`;

Container.AuthForm = AuthForm;

export default Container;