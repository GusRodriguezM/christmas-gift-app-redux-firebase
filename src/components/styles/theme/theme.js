const theme = {
    colors: {
        black: '#000000',
        loginButton: '#4285f4',
        disabledButton: '#999595',
        red: '#C30F16',
        white: '#FFFFFF',
        list: '#c8e4e4',
        navbar: '#e8ceac',
        boxShadow: 'rgba(9, 9, 9, 0.23)',
    }
}

export default theme;