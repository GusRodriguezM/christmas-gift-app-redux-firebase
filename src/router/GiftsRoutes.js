import React from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import { GiftScreen } from '../components/gifts/GiftScreen';
import { Navbar } from '../components/navbar/Navbar';

export const GiftsRoutes = () => {
    return (
        <>
            <Navbar />

            <Routes>
                <Route path="gifts" element={<GiftScreen />} />

                <Route path="/*" element={<Navigate to="/gifts"/> } />
            </Routes>
        </>
    )
}
