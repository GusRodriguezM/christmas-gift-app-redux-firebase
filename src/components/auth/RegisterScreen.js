import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { login } from '../../store/slices/auth/authSlice';
import { useForm } from '../hooks/useForm';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import { Button } from '../styles/shared/Button.styled';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const navigate = useNavigate();

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: ''
    });

    const { name, email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        
        dispatch( login(email) );
        navigate('/', {
            replace: true
        });
    }

    const handleNavigate = () => {
        navigate('/auth/login', {
            replace: true
        });
    }

    return (
        <Container>

            <h1 className='login__title'>Sign in</h1>

            <Container.AuthForm onSubmit={handleLogin} className='login__form'>

                <Input
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='login__form--input'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />

                <Input 
                    type='text'
                    placeholder='Email'
                    name='email'
                    className='login__form--input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='login__form--input'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button type='submit'>
                    <span>Login</span>
                </Button>

            </Container.AuthForm>

            <Container.Span>I already have an account</Container.Span>
            <Button
                onClick={handleNavigate}
            >
                Login
            </Button>
        </Container>
    )
}
