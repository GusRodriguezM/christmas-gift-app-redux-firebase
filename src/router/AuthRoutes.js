import React from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import { LoginScreen } from '../components/auth/LoginScreen';
import { RegisterScreen } from '../components/auth/RegisterScreen';

export const AuthRoutes = () => {
    return (
        <Routes>
            <Route path="login" element={<LoginScreen />} />
            <Route path="register" element={<RegisterScreen />} />

            <Route path="/*" element={<Navigate to="/auth/login" />} />
        </Routes>
    )
}
