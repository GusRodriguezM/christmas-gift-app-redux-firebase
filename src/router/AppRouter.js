import React from 'react';
import { Route, Routes } from 'react-router-dom';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';
import { AuthRoutes } from './AuthRoutes';
import { GiftsRoutes } from './GiftsRoutes';
import { useCheckAuth } from '../hooks';

export const AppRouter = () => {

    const status = useCheckAuth();

    //TODO: Make a component (spinner) to show the app is loading
    if(status === 'checking') 
        return (<h1>Loading...</h1>)

    return (
        <>
            <Routes>

                <Route path="auth/*" element={
                    <PublicRoute>
                        <Routes>
                            <Route path="/*" element={<AuthRoutes />} />
                        </Routes>
                    </PublicRoute>
                } />

                <Route path="/*" element={
                    <PrivateRoute>
                        <GiftsRoutes />
                    </PrivateRoute>
                } />

            </Routes>

        </>
    )
}