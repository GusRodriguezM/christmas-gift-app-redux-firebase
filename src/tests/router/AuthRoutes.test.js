import { configureStore } from "@reduxjs/toolkit";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import theme from "../../components/styles/theme/theme";
import { AuthRoutes } from "../../router/AuthRoutes";
import { authSlice } from "../../store/slices/auth";
import { uiSlice } from "../../store/slices/ui";
import { nonAuthenticatedState } from "../fixtures/authFixtures";

describe('Tests in the <AuthRoutes />', () => {

    test('Should render the component correctly', () => {

        //Mocking the store
        const store = configureStore({
            reducer: {
                auth: authSlice.reducer,
                ui: uiSlice.reducer,
            },
            preloadedState: {
                auth: nonAuthenticatedState
            }
        });

        //Rendering the component
        const { container } = render(
            <Provider store={ store }>
                <MemoryRouter initialEntries={['/login']}>
                    <ThemeProvider theme={ theme }>
                        <AuthRoutes />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Expecting that the component rendered matches with the snapshot
        expect( container ).toMatchSnapshot();

        //Expecting that the component have the elements for an input, link, heading and button
        expect( screen.getAllByText('Login').length ).toBeGreaterThanOrEqual(1);
        expect( screen.getAllByRole('textbox').length ).toBe(1);
        expect( screen.getAllByRole('link').length ).toBe(1);
        expect( screen.getAllByRole('button').length ).toBe(2);

    });

});