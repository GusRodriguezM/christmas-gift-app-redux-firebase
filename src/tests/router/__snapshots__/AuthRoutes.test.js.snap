// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in the <AuthRoutes /> Should render the component correctly 1`] = `
<div>
  <div
    class="sc-hKMtZM bGngom"
  >
    <h1>
      Login
    </h1>
    <form
      aria-label="login-form"
      class="sc-dkzDqf bnSFoi"
    >
      <input
        autocomplete="off"
        class="sc-eCYdqJ drExBT"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="sc-eCYdqJ drExBT"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="sc-bczRLJ glPbkZ"
        type="submit"
      >
        <span>
          Login
        </span>
        <i
          class="fa-solid fa-user-lock"
        />
      </button>
      <button
        aria-label="googleBtn"
        class="sc-bczRLJ glPbkZ"
      >
        <span>
          Sign in with Google
        </span>
        <i
          class="fa-brands fa-google"
        />
      </button>
    </form>
    <a
      aria-disabled="false"
      href="/auth/register"
    >
      Create a new account
    </a>
  </div>
</div>
`;
